app/lib/components/scrollbar/Scrollbar.tsx

```diff
@@ -0,0 +1,170 @@
+    this.thumbX =
+      scrollLeftMax === 0
+        ? 0
+        : (this.scrollLeft / this.scrollLeftMax) * (trackClientWidth - this.thumbWidth);
+
+    this.thumb!.style.left = `${this.thumbX}px`;
+    this.thumb!.style.width = `${this.thumbWidth}px`;
+
+    this.track.setAttribute(
+      "aria-valuenow",
+      String(
+        Math.round(
+          this.thumbX === 0 ? 0 : (100 * this.thumbX) / (trackClientWidth - this.thumbWidth)
+        )
+      )
+    );
+  }
+
+  private isScrolling = false;
+  // Single scroll event here avoids creating too many objects.
+  private scrollEvent = { delta: 0, animate: false };
+  private minMouseX = 0;
+  private maxMouseX = 0;
+  private mouseX = 0;
+  onMouseDown = (e: MouseEvent) => {
+    if (e.buttons & 1) {
+      this.isScrolling = true;
+      const thumb = this.thumb!.getBoundingClientRect();
+      const track = this.track!.getBoundingClientRect();
+      this.minMouseX = e.clientX - (thumb.x - track.x);
+      this.maxMouseX = e.clientX + (track.x + track.width - (thumb.x + thumb.width));
+    }
+    this.updateMouse(e);
+  };
+  onMouseMove = (e: MouseEvent) => {
+    if (!this.isScrolling) return;
+    if (!(e.buttons & 1)) {
+      this.isScrolling = false;
+      return;
+    }
+
+    const lastX = this.mouseX;
+    this.updateMouse(e);
+    const deltaX = this.mouseX - lastX;
+
+    if (deltaX !== 0 && this.track!.clientWidth - this.thumbWidth !== 0) {
+      this.scrollEvent.delta =
+        (deltaX / (this.track!.clientWidth - this.thumbWidth)) * this.scrollLeftMax;
+      this.scrollEvent.animate = false;
+      this.events.dispatch("scroll", this.scrollEvent);
+    }
+  };
+  onMouseUp = (e: MouseEvent) => {
+    this.updateMouse(e);
+    this.isScrolling = false;
+  };
+  updateMouse = (e: MouseEvent) => {
+    this.mouseX = Math.min(this.maxMouseX, Math.max(this.minMouseX, e.clientX));
+  };
+
+  private isMouseInside = false;
+  onMouseEnterScrollingElement = (e: MouseEvent) => {
+    this.isMouseInside = true;
+  };
+  onMouseOutScrollingElement = (e: MouseEvent) => {
+    this.isMouseInside = false;
+  };
+  onWheelScrollingElement = (e: WheelEvent) => {
+    if (e.shiftKey) {
+      e.preventDefault();
+      this.scrollEvent.delta = e.deltaY;
+      this.scrollEvent.animate = true;
+      this.events.dispatch("scroll", this.scrollEvent);
+    }
+  };
+  onKeyDown = (e: KeyboardEvent) => {
+    if (!this.isMouseInside || e.ctrlKey || e.shiftKey) return;
+
+    if (e.which === 39 || e.which === 37) {
+      e.preventDefault();
+      const dir = e.which - 38;
+      this.scrollEvent.delta = 40 * dir;
+      this.scrollEvent.animate = true;
+      this.events.dispatch("scroll", this.scrollEvent);
+    }
+  };
+
+  dispose() {
+    this.disposer.dispose();
+  }
+}
+
+export type HorizontalScrollbarProps = {
+  setController: (controller: HorizontalScrollbarController) => void;
+};
+
+export function HorizontalScrollbar({ setController }: HorizontalScrollbarProps) {
```

Comment:
can we make this a class instead of a function?

maybe HorizontalScrollbarComponent to match our other components